### PR TITLE
Converted Configuration to TypeScript

### PR DESCRIPTION
**Pull request by sebjulliand**
### Disclaimer
This is my first big PR here, and my take on this conversion may or may not fit the project. I'm all open to discuss it and see it rejected 😊

### Changes
* Converted Configuration to TypeScript
* Used namespaces to differentiate Global configuration from Connections'
* Proposed Configuration interfaces hierarchy
* Added type JSDOC where relevant

### Checklist
* [x] have tested my change
* [x] updated relevant documentation
* [x] eslint is not complaining
* [x] have added myself to the contributors' list in [CONTRIBUTING.md](https://github.com/halcyon-tech/vscode-ibmi/blob/master/CONTRIBUTING.md)

### CODE REVIEW

**Comment by worksofliam** [comment]
@sebjulliand woah, you weren't kidding!

I will be reviewing this either later tonight or tomorrow morning.

Thanks for your great effort on this front.

**Comment by sebjulliand** [comment]
Happy to help! I'll be glad to convert more whenever I can.

**Comment by worksofliam** [comment]
@sebjulliand I fixed a conflict here and now I am going to review.

File: src/api/IBMi.ts [hunk 1]
```diff
@@ -602,6 +604,9 @@ export default class IBMi {
         error: e
       };
     }
+    finally{
+      ConnectionConfiguration.update(this.config!);
```
Comment (worksofliam) [inline src/api/IBMi.ts:608]: This is a nice change.

File: src/webviews/settings/index.js [hunk 2]
```diff
@@ -193,7 +194,7 @@ module.exports = class SettingsUI {
         field = new Field(`submit`, `save`, `Save settings`);
         ui.addField(field);
     
-        let {panel, data} = await ui.loadPage(`Settings: ${name}`);
+        let {panel, data} = await ui.loadPage(`ConnectionConfiguration: ${name}`);
```
Comment (worksofliam) [inline src/webviews/settings/index.js:197]: This should still say `Settings: ${name}` as this is what appears on the tab.

File: src/webviews/settings/index.js [hunk 3]
```diff
@@ -279,7 +281,7 @@ module.exports = class SettingsUI {
 
           ui.addField(new Field(`submit`, `submitButton`, `Save`));
 
-          const {panel, data} = await ui.loadPage(`Login Settings: ${name}`);
+          const {panel, data} = await ui.loadPage(`Login ConnectionConfiguration: ${name}`);
```
Comment (worksofliam) [inline src/webviews/settings/index.js:284]: This should also still say `Settings`

**Review by worksofliam — CHANGES_REQUESTED** [review]
First off, let me say - this is an amazing PR. I really appreciate the effort of moving some of the most complicated code (the configuration setup is a lot!) into TS.

I very much like the seperation of `GlobalConfiguration` and `ConnectionConfiguration` namespaces.

Other than the two issues I've left comments for, it's all looking good. It just needs some user testing as well.

**Comment by worksofliam** [comment]
It's working very well for me. Even debugged a little bit to try out the new config APIs.

@halcyon-tech/core Any chance we could do some user testing on this to make sure there is no impact of functionality?

Thanks!

File: src/webviews/settings/index.js (hunk 2 above)
Comment (sebjulliand) [reply src/webviews/settings/index.js:197]: Eek, that was a **Replace All** that went wrong 😅

File: src/webviews/settings/index.js (hunk 3 above)
Comment (sebjulliand) [reply src/webviews/settings/index.js:284]: Same mistake as above.

**Comment by sebjulliand** [comment]
> First off, let me say - this is an amazing PR. I really appreciate the effort of moving some of the most complicated code (the configuration setup is a lot!) into TS.
> 
> I very much like the seperation of `GlobalConfiguration` and `ConnectionConfiguration` namespaces.
> 
> Other than the two issues I've left comments for, it's all looking good. It just needs some user testing as well.

Much obliged @worksofliam ! I'm glad to finally have found the time to participate and give back to the community by helping with this great project.

**Review by worksofliam — APPROVED** [review]
Code is good. Further user testing coming.

**Comment by Wright4i** [comment]
@worksofliam I'll take a look tonight and give my feedback!

**Comment by worksofliam** [comment]
@Wright4i Any luck with the user testing at all?

**Comment by Wright4i** [comment]
@worksofliam so far so good! I actually had a bit of work to do this morning on some programs so I've put in a bit of real world testing as well. No issues to report. I'll reach out if something pops up but I think this is good to go.

**Comment by worksofliam** [comment]
@Wright4i Sorry to pester you, but very glad to hear it's looking good! Will merge this today.

**Pull request merged**